### DISCUSSION TITLE
Netbox Migration

### DISCUSSION THREAD

**Discussion by Snowbull08** [netbox-community/netbox#11253] [Q&A]
Hi, I am new to Netbox, and in my new Job, we have Netbox 2.1.5 installed on Ubuntu 16. Notes from my predecessor advise that all upgrade paths weren't fruitful. Please suggest the migration path from 2.1.5 to the latest version.

**Comment by mtinberg** [comment]
You will probably need to read through the release notes for each major version from 2.1 to 2.11 to 3.4 as there may be steps you need to complete, eg. upgrade to a current version of PostgreSQL, but the major one is that 3.x doesn't include the db migration scripts from 2.x prior to 2.11.x, so you have to do a stepwise upgrade to 2.11.x first, and then you can upgrade to current.  Other people who've had to refresh an older instance like this have shared some notes in this mailing list which you may be able to search for, common problems are that pip install will grab the latest of everything, but that often installs versions of dependencies than the old version of Netbox was tested against and isn't compatible with, so you have to pin an older version in requirements.txt.

I would start by using the PostgreSQL tools (pg_dump) to dump the database on the old system, then install 2.11.x on a new system, get that working, then restore the old 2.1 database onto that new system and run the upgrade.sh script as documented for any upgrade.  Then you can upgrade the new system to the latest Netbox and run the upgrade procedure again to get a new system with a new 3.4.x database.  If you run into any problems, you can always start over and try again, as you haven't broken the existing Netbox 2.1 install you are relying on.


—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________
From: Snowbull08 ***@***.***>
Sent: Monday, December 19, 2022 11:43 PM
To: netbox-community/netbox ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [netbox-community/netbox] Netbox Migration (Discussion #11253)


Hi, I am new to Netbox, and in my new Job, we have Netbox 2.1.5 installed on Ubuntu 16. Notes from my predecessor advise that all upgrade paths weren't fruitful. Please suggest the migration path from 2.1.5 to the latest version.

**Reply by markkuleinio** [reply]
> common problems are that pip install will grab the latest of everything, but that often installs versions of dependencies than the old version of Netbox was tested against and isn't compatible with, so you have to pin an older version in requirements.txt.

FWIW, I set up https://github.com/markkuleinio/netbox-freezes if that helps someone at some point. It only starts at NetBox 3.1.8 though.

**Comment by jeremystretch** [comment]
Please consult the [official documentation](https://docs.netbox.dev/en/stable/installation/upgrading/) rather than relying on your predecessor's notes. If you abide by the upgrade instructions carefully you're unlikely to encounter any issues.

**Reply by Snowbull08** [reply]
I followed the document and did encounter some issues: I tried Python 3.7, 3.9.1 and always get stuck at

 File "/opt/netbox-2.11.10/venv/lib/python3.7/site-packages/rest_framework/utils/encoders.py", line 14, in <module>
    from rest_framework.compat import coreapi
  File "/opt/netbox-2.11.10/venv/lib/python3.7/site-packages/rest_framework/compat.py", line 71, in <module>
    import markdown
  File "/opt/netbox-2.11.10/venv/lib/python3.7/site-packages/markdown/__init__.py", line 29, in <module>
    from .core import Markdown, markdown, markdownFromFile  # noqa: E402
  File "/opt/netbox-2.11.10/venv/lib/python3.7/site-packages/markdown/core.py", line 26, in <module>
    from . import util
  File "/opt/netbox-2.11.10/venv/lib/python3.7/site-packages/markdown/util.py", line 87, in <module>
    INSTALLED_EXTENSIONS = metadata.entry_points().get('markdown.extensions', ())
AttributeError: 'EntryPoints' object has no attribute 'get'

**Reply by Snowbull08** [reply]
What do it do from here

**Reply by joaolucasmacedo** [reply]
Use Python 3.6 and run upgrade by each major release until 2.11.12.
2.1 >> 2.x >> 2.11
Than install Python 3.8 and run the upgrade to 3.0 >>3.1 >> 3.2.
Than install Python 3.10 and run the upgrade to 3.3 >> 3.4.
Note that on the upgrade to 3.2 and to 3.4 there are some field you need to empty.

I've been using Netbox since 2.6 and I can tell you that these are the best Python version for each case.
Although you could use different Python versions and jump from a "old" release to a new one, in my experience I've seen many problems. So, the safest way is to run each upgrade moving to major releases one by one and considering that Python versions.

Anyway, you have a lot of work to do, because many things have changed since Netbox 2.1, like the Python virtual environment, systemd services and the data models. Maybe the best approach for your case is to make a fresh install and migrate data manually using CSV files. It is also a good opportunity to review your data and the way it is modeled on Netbox.

**Reply by candlerb** [reply]
> Use Python 3.6 and run upgrade by each major release until 2.11.12

I disagree: there is no need to go via intermediate releases.  In principle you can go straight from 2.1.x to 2.11.12, then one more upgrade to 3.4.1.  (You *may* find an issue where you're using legacy contact fields which need migrating, in which case, it's *possible* an intermediate version between 2.11.12 and 3.4.1 is required.  But if this is true, `./upgrade.sh` will tell you)

Python versions 3.6, 3.7 or 3.8 are fine for Netbox v2.11.12.  Python versions 3.8, 3.9, 3.10 or 3.11 are fine for Netbox v3.4.1.  So using python3.8 will cover all cases. Otherwise, use any of python3.6-3.8 to get to netbox v2.11.12, dump the SQL database, then install netbox v3.4.1 with any of python3.8-3.11 and restore the database.

The problem @Snowbull08 has is different.

There is a known potential problem here: Netbox installs are not generally reproducible, because dependency versions are not pinned.  That is: whilst it was possible to install v2.11.12 a year ago, it is not necessarily possible to install it today, because in the mean time, newer (and incompatible) versions of the libraries that Netbox depends on have been released.

However, I've tried reproducing this problem, by doing a fresh install of Netbox v2.11.12 under Ubuntu 20.04 (which has python3 version 3.8.10), and it succeeded without problems.

Therefore, the problem is something else with @Snowbull08's environment.  Can you say what OS you're trying to install under? Where did these version of python 3.7 and 3.9.1 come from?  And most importantly, are you using the `upgrade.sh` script to perform the upgrade?  This is critical, because it builds a python virtualenv.

If you're trying to install with a version of python which is installed as anything other than `python3`, then you need to set an environment variable: e.g.

```
PYTHON=python3.8 ./upgrade.sh
```

And note that you definitely *can't* use python3.9 with Netbox v2.11.12.

> Maybe the best approach for your case is to make a fresh install and migrate data manually using CSV files.

Sorry, but I *very strongly* disagree with this. It is impossible to make an accurate export/import of data using CSV files in Netbox, for a whole range of reasons that I won't go into here.

But I *do* agree that you should make a fresh install of v2.11.12, say in a new VM.  Import the SQL database dump from the old system and run upgrade.sh, as @mtinberg described (plus any media files) - also documented in [replicating netbox](https://docs.netbox.dev/en/stable/administration/replicating-netbox/).  Then repeat for the upgrade to v3.4.1.

**Reply by joaolucasmacedo** [reply]
I'll not disagree with you @candlerb , but that's the safest way I often recommend to someone who is not so familiar with Netbox environments. It is more about doing things piece by piece.

**Reply by julianze** [reply]
> What do it do from here

@Snowbull08 
I had the same yesterday with v2.11.10.

The package importlib_metadata has a major update and this has broken netbox.

Pin a version around the date of your installation or the last successful upgrade in your requirements.txt:
https://github.com/python/importlib_metadata/releases

pip3 delivered 6.0.0 the last days. I had to go back to 4.6.1.
Maybe you can identify your last working importlib_metadata in your logs, cache, history etc

**Reply by candlerb** [reply]
FWIW, this is from a test 2.11.12 system that I installed a few days ago (Jan 3):

```
(venv) root@netbox-test:/opt/netbox# pip freeze | grep -i importlib
importlib-resources==5.10.2
(venv) root@netbox-test:/opt/netbox#
```

There is no `importlib[-_]metadata` present. I wonder if some other package has also updated, and is now pulling this in?  Or maybe a plugin that you're using?

You can get the full `pip freeze` output from my system at https://github.com/markkuleinio/netbox-freezes/blob/main/netbox-2.11.12.txt (thanks @markkuleinio for starting this repository!)

**Reply by julianze** [reply]
It's a dependency of Markdown, which is pinned:
```
(venv) root@netbox:/opt/netbox# pip3 show importlib-metadata                       Name: importlib-metadata
Version: 4.6.1
Summary: Read metadata from Python packages
Home-page: https://github.com/python/importlib_metadata
Author: Jason R. Coombs
Author-email: jaraco@jaraco.com
License: UNKNOWN
Location: /opt/netbox/venv/lib/python3.7/site-packages
Requires: typing-extensions, zipp
Required-by: click, jsonschema, Markdown, redis
(venv) root@netbox:/opt/netbox#
```

**Reply by candlerb** [reply]
Hmm. Are we both talking about 2.11.12?  I have importlib-metadata on 3.4.2 but not 2.11.12.

On my 2.11.12 test system:

```
(venv) root@netbox-test:/opt/netbox# pip freeze | egrep -i 'markdown|importlib'
importlib-resources==5.10.2
Markdown==3.3.4
```

On 3.4.2:

```
(venv) root@netbox3:/opt/netbox# pip freeze | egrep -i 'markdown|importlib'
importlib-metadata==4.13.0
importlib-resources==5.10.2
Markdown==3.3.7
```

**Reply by candlerb** [reply]
Ah I think I see: I'm testing with Ubuntu 20.04 / python 3.8.  The dependency on importlib-metadata is only for python < 3.8.

```
root@netbox-test:/opt/netbox# grep Requires venv/lib/python3.8/site-packages/Markdown-3.3.4.dist-info/METADATA
Requires-Python: >=3.6
Requires-Dist: importlib-metadata ; python_version < "3.8"
Requires-Dist: coverage ; extra == 'testing'
Requires-Dist: pyyaml ; extra == 'testing'
```

**Reply by julianze** [reply]
I have a netbox v2.11.10 on this system:
```
root@netbox:~# python3 --version
Python 3.7.3
root@netbox:~# cat /etc/debian_version
10.9
```

**Reply by candlerb** [reply]
OK, so that's it: with python 3.7, Markdown requires importlib-metadata (any version); with python 3.8 it does not.

Therefore, another solution would be to upgrade your base python version to 3.8.

And/or you could report the dependency problem with Markdown under python 3.7 upstream.  (However 3.7 is not too far from [end of life](https://devguide.python.org/versions/) - 27 Jun 2023)
